protocol.go

```diff
@@ -80,9 +69,14 @@ func NewConn(conn net.Conn, timeout time.Duration) *Conn {
 // it is returned and the socket is closed.
 func (p *Conn) Read(b []byte) (int, error) {
 	var err error
-	p.once.Do(func() { err = p.checkPrefix() })
+	p.once.Do(func() { err = p.checkHeader() })
```

Comment:
I like your changes so far. Would this violates the spec? http://www.haproxy.org/download/1.5/doc/proxy-protocol.txt

> The receiver MUST be configured to only receive the protocol described in this
specification and MUST not try to guess whether the protocol header is present
or not. This means that the protocol explicitly prevents port sharing between
public and private access. Otherwise it would open a major security breach by
allowing untrusted parties to spoof their connection addresses. The receiver
SHOULD ensure proper access filtering so that only trusted proxies are allowed
to use this protocol.